File: RDParser.cpp

//Aalap Shastri
//Project 2    CS280

#include <iostream>
#include <fstream>
#include <string>
using namespace std;

int value;
string app;

bool A(void);
bool I(void);
bool E(void);
bool P(void);
bool O(void);
bool L(void);
bool U(void);
bool C(void);
bool D(void);






int main(){
    
    cout << "A -> I = E |E" << app << endl;
    cout << "E -> P O P | P" << app << endl;
    cout << "O -> + | - | * | / | **" << app << endl;
    cout << "P -> I | L | UI | UL | (E)" << app << endl;
    cout << "U -> + | - | !" << app << endl;
    cout << "I -> C | CI" << app << endl;
    cout << "C -> a | b | ... | y | z " << app << endl;
    cout << "L -> D | DL" << app << endl;
    cout << "D -> 0 | 1 | ... | 8 | 9 " << app << endl;
    cout << " " << app << endl;
    
        ifstream document ("input.txt");
    
        if (document.is_open())
        {
            while (!document.eof())
            {
                value = 0;
                getline(document, app);
                cout << "String that was read from text file: " << app << endl;
                
                
                if (A())
                {
            cout << "The string's in the language\n" << endl;
                }
                else
                {
            cout << "The string is not in the language\n" << endl;
                }
            }
            document.close();
            }
        return 0;
       }

//***************************

      bool A()
      {
       if (I())
       {
        if (value < app.length() && app[value] == '=')
        {
            ++value;
            if (E()){
                    return true;
                      }
                    }
                  }
                 return false;
                  }

//***************************

        bool E()
        {
         if (P())
         {
             ++value;
             if(O())
             {
                 return true;
             }
             
             else if(P())
             {
                 return true;
             }}
            return false;
             {
                 
             }
        }

            
            //***************************
             

            bool O()
            {
                if(app[value] == '+' || app[value] == '-' || app[value] == '*' || app[value] == '/' || app[value] == '**'){
                
                    ++value;
                    return true;
                }
                 return false;
                }



             bool P()
             {
                 if (I())
                 {
                     return true;
                 }
                 else if (L())
                 {
                     return true;
                 }
                 else if(U())
                 {
                     ++value;
                     if(I())
                     {
                         return true;
                     }
                     else if(L())
                     {
                         return true;
                     }
                     else
                     {
                         return false;
                     }
                 }
                 else
                 {
                     if (value < app.length() && app[value] == '('     )
                     {
                         ++value;
                         if (E())
                         {
                             if (value < app.length() && app[value] == ')'      )
                             {
                                 ++value;
                                 return true;
                             }
                         }
                     }
                 }
                 return false;
             }


             bool U()
             {
                 if(app[value] == '+' || app[value] == '-'  || app[value] == '!')
                 {
                     ++value;
                     return true;
                 }
                 return false;
             }
             

             bool I()
             {
                 if(C())
                 {
                     ++value;
                 }
                 else if(I())
                 {
                     return true;
                 }
                 else
                 {
                     return true;
                 }
                 return false;
             }


             
             bool C()
             {
                 if (value < app.length() && (app[value] >= 'a' && app[value] <= 'z'))
                 {
                     ++value;
                     return true;
                 }
                 return false;
                 
             }

             bool L()
             {
                 if(D())
                 {
                     ++value;
                 }
                 
                 else if(L())
                 {
                     return true;
                 }
                 else
                 {
                     return true;
                 }
                 return false;
             }
             


             bool D()
             {
                 if (value < app.length() && (app[value] >= '0' && app[value] <= '9'))
                 {
                     ++value;
                     return true;
                 }
                 return false;
             }




             
             
             
             
             
             
             
             
             
             



        







         

























   

File: RDParser.java




import java.io.*;
import java.util.*;
import java.util.Scanner;
import java.lang.String;

public class RDParser {
	
	public static Scanner scanner;  
	
	public static int value;
	public static String app;
	
	
	public static void main(String[] args) {
	
		System.out.println("A -> I = E |E"); // 
		System.out.println("E -> P O P | P"); //
		System.out.println("O -> + | - | * | / | **");   //
		System.out.println("P -> I | L | UI | UL | (E)");//
		System.out.println("U -> + | - | !"); // 
		System.out.println("I -> C | CI"); 
		System.out.println("C -> a | b | ... | y | z ");
		System.out.println("L -> D | DL");
		System.out.println("D -> 0 | 1 | ... | 8 | 9 ");
		
		System.out.println(" ");   
		
		String filer;
		
		RDParser document = new RDParser();
		filer = "input.txt";
		if(document.open(filer)== 0)
		{
			while(scanner.hasNextLine())
			{  
				value = 0;  
				document.reader();
				}
			scanner.close(); 
			}
		}
	public int open(String Filer) 
	{ 
		try
		{
	     scanner = new Scanner(new File(Filer));  
	     return 0;  
	     }  
		catch(FileNotFoundException x)  
		{    System.out.println("No file has been found"); 
		return 1;}
		}
	
		
	public void reader() 
	{
		app = scanner.nextLine(); 
		System.out.println("String read from file :" + app);
		if(A() && value == app.length())
		{   
			System.out.println("The String:   " + app + "  Is In The Language!");
			}
		else{  System.out.println("The String: " + app + "  Is Not In The Language!");
		} 
		}
	
	
	
	private static boolean I() {
		return false;
	}
	private static boolean E() {
		return false;
	}
	private static boolean P() {
		return false;
	}
	private static boolean O(){
		return false;
	}
	private static boolean L() {
		return false;
	}
	private static boolean U() {
		return false;
	}
	private static boolean C() {
		return false;
	}
	private static boolean D() {
		return false;
	}
	
	public static boolean A() 
	{
		if (I())
		{
			if (value < app.length() && app.charAt(value) == '=') 
			{
				value++;
				if (E())
				{
					return true;
					}
				}
			}
		return false;
		
		//**********************************************
		

        public static boolean E()
        {
        	if (P())
         {
             value++;
             if(O())
             {
                 return true;
             }
             
             else if(P())
             {
                 return true;
             }}
            return false;
             {
                 
             }
        }
      //**********************************************
		
        public static boolean O()
        {
            if(app.CharAt(value) == '+' || app.CharAt(value) == '-' || app.CharAt(value) == '*' || app.CharAt(value) == '/' || app.CharAt(value) == '**'){
            
                value++;
                return true;
            }
             return false;
        }
        
        //**********************************************
        
        public static boolean P()
        {
        	if (I())
        	{
        		return true;
        		}
        	else if
        	(L())
        	{
        		return true;
        		}
        	else if(U())
        	{
        		value++;
        		if(I())
        		{
        			return true;
        			}
                else if(L())
                {
                	return true;  
                	}
                else{        
                	return false;     
                	} 
        		}
        	else
        	{
        		if (value < app.length() && app.charAt(value) == '('    ) 
        		{
        			value++;
        			if (E())      
        			{
        				if (value < app.length() && app.charAt(value) == ')'   )
        				{            
        					value++;
        					return true;
        					}
        				}
        			}
        		}
        		return false;
        }
        
        //*************************************
        		
        public static boolean U()
        {
        	if (app.charAt(value) == '+' || app.charAt(value) == '-' || app.charAt(value) == '!') 
        	{
        		value++;
        		return true;
        		}
        	return false;
        	}
      //*************************************
        public static boolean I() 
        {
        	if (C()){return true;}
        	else if(C())
        	{
        		value++;
        		if(I())
        		{
        			return true;
        			}
        		return false;
        		} 
        	return false;
        	}
        
      //*************************************		
        public static boolean C()
        {
        	if (value < app.length() && (app.charAt(value) >= 'a' && app.charAt(value) <= 'z')) 
        	{
        		value++;
        		return true; 
        		} 
        	return false;
        }
      //*************************************		
        
        public static boolean L()
        {
        	if (D()){return true;
        	}
        	else if(D())
        	{
        		value++;
        		if(L())
        		{
        			return true;
        			}
        		return false;
        		}
        	return false;
        	}
      //*************************************
        
        public static boolean D()
        {
        	if (value < app.length() && (app.charAt(value) >= '0' && app.charAt(value) <= '9'))
        	{
        		value++;
        		return true;
        		}
        	return false;
        
        
        
        	}
	}
	

}
